### PR TITLE
add support for building first ARM-based debian package

### PR DESCRIPTION
**Pull request by Govinda-Fichtner**
This PR adds support for building debian packages for ARM and is part of the work on issue #17802.

As a first step we have added support for the jessie version of the debian distribution.
Later PR's will add support for more distribution and version combinations.

Initially I refrained from adding a generate.sh/build.sh bash script for arm/deb as it only adds complexity. If this is necessary later to keeps things DRY I plan to add these as a separate PR.

Most of this PR is about making the affected scripts aware of the architecture they are running on.

I've been testing the changes on ARM for `make deb` and on AMD64 for `make deb` and `make rpm` to ensure that nothing got broken by the changes.
Still there were some areas which I have not been able to tests properly - e.g. in contrib/reprepro.

### CODE REVIEW

**Comment by thaJeztah** [comment]
@Govinda-Fichtner looks like this needs a rebase

**Comment by Govinda-Fichtner** [comment]
@thaJeztah This one slipped through. Rebased again.

**Comment by Govinda-Fichtner** [comment]
@thaJeztah Updated PR to be in sync with master.

File: contrib/builder/arm/deb/debian-jessie/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM ioft/armhf-debian:jessie
```
Comment (thaJeztah) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: wondering; I know this is a temporary solution, but can we / should we use a more official image for this? At least something under the `docker/` namespace?

ping @tianon for ideas/suggestions

File: contrib/builder/arm/deb/debian-jessie/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+RUN apt-get update && apt-get install -y apparmor bash-completion btrfs-tools build-essential curl ca-certificates debhelper dh-apparmor dh-systemd git libapparmor-dev libdevmapper-dev libltdl-dev libsqlite3-dev libsystemd-journal-dev --no-install-recommends && rm -rf /var/lib/apt/lists/*
+
+ENV GO_VERSION 1.5.1
+RUN curl -fSL "https://github.com/hypriot/golang-armbuilds/releases/download/v${GO_VERSION}/go${GO_VERSION}.linux-armv7.tar.gz" | tar xzC /usr/local
```
Comment (thaJeztah) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: Possibly the same here, I know there's been discussion around using binaries in stead of building from source. Not sure what our options are here, so just commenting :smile:

File: contrib/builder/arm/deb/debian-jessie/Dockerfile (hunk 1 above)
Comment (Govinda-Fichtner) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: @thaJeztah Couple of weeks ago I've been asking @jfrazelle  if it would make sense to a least move all the necessary images for building and testing Docker on ARM to the official Docker image library. Back then she said that for the time being we should just use the private images. 

I guess @jfrazelle meant that we should clean up the images later as a follow up task. 

In my opinion this is really about creating an official Docker image library for ARM, which I really would like to see sooner than later. Still I am not sure if this should be something we solve as part of this PR.

Comment (thaJeztah) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: @Govinda-Fichtner alright, seems fair. I just noticed it here. I'm good with keeping it separate from this PR

File: contrib/builder/arm/deb/debian-jessie/Dockerfile (hunk 2 above)
Comment (DieterReuter) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: @thaJeztah to speed up the build process we'd recommend using pre-compiled Go versions, and it's a lot easier than building Go 1.5.1 from source too. There is already an open issue at https://github.com/golang/go/issues/13066 and we're expecting to get officially built pre-compiled Go tarballs for ARM later, but this is scheduled for Go 1.6. This been said, we'd like to take the easy and fast approach and using our own pre-compiled versions of Go for ARMv7. BTW, this build process is already documented in that specific repo. And as soon as the official Go tarballs for ARM are available, we'll change this to the official downloads.

**Comment by tianon** [comment]
IMO it's still too soon to be worrying about `.deb` file generation (that's a "release ARM support" concern, when we've just barely merged "ARM support" and are still working out the kinks in that)

**Comment by Govinda-Fichtner** [comment]
@tianon Hmm, not sure if I understood what you meant.

With the recently merged PR's for ARM it is now possible to run make with targets such as `binary`, `test-unit`, `test-integration-cli`. All unit and integration tests should work now with the exception of 3 integration tests out of ~ 900 integration tests in total. 

Thus being able to create debian packages seems to be the next logical step. Looking at the `Makefile` at the root of the docker/docker repository I see no other major make targets that need to be addressed.
Am I missing something?

**Comment by tianon** [comment]
Do we have CI set up yet to ensure the behaviour doesn't regress (or at
least that we catch it quickly)?  IMO, building debs shows a level of
support commitment we aren't at yet.

**Comment by Govinda-Fichtner** [comment]
@tianon \cc @jfrazelle My understanding is that all the current work around #17802 (Support Docker on ARM) is done to be able to set up a CI process for ARM at Docker as soon as possible. Being able to build a deb package closes a first slice that makes it possible to build, test and package on ARM. 
This slice needs to be expanded with further PR's down the road to support more distributions and package formats, but the basic are there.

At Hypriot we are already running all the supported make targets in our Jenkins-based CI environment.

How would you like to handle this?

**Comment by thaJeztah** [comment]
ping @jfrazelle I assigned you to this one, can you have a look

**Comment by thaJeztah** [comment]
@Govinda-Fichtner looks like this needs a rebase :cry:

**Comment by jfrazelle** [comment]
going to try this out today thanks for your patience

**Comment by jfrazelle** [comment]
can we fix this stuff first too https://jenkins.dockerproject.org/job/Docker%20Master%20(arm)/112/console

**Comment by Govinda-Fichtner** [comment]
@jfrazelle /cc @StefanScherer That is because of the many changes to the Dockerfile in the last weeks (since this PR was created). For this specific error I guess https://github.com/docker/docker/commit/359d0c247fe523ef82676f9ae4046b79e6854fe1 is missing here.

**Comment by Govinda-Fichtner** [comment]
@jfrazelle I created a new [PR](https://github.com/docker/docker/pull/18728) which should fix the your problems from the ARM builder.

**Comment by thaJeztah** [comment]
ping @jfrazelle what's the status on this one? does this depend on https://github.com/docker/docker/issues/18728?

**Comment by thaJeztah** [comment]
https://github.com/docker/docker/pull/18728 was just merged

**Comment by jfrazelle** [comment]
looks like its just needs a rebase :)

**Comment by StefanScherer** [comment]
Sure ;-)

File: contrib/builder/arm/deb/debian-jessie/Dockerfile (hunk 1 above)
Comment (tianon) [inline contrib/builder/arm/deb/debian-jessie/Dockerfile]: I've got that part in progress/mostly complete now for armhf, plc64le, and s390x. :+1: (will be making a PR for that part soonish)

**Comment by tianon** [comment]
Instead of just bare "arm", can we use "armhf"? (see also https://github.com/docker/docker/pull/18987)

**Comment by tianon** [comment]
Also, I'd rather this use "builder/deb/ARCH" since all the deb Dockerfiles should be similar enough to be generated at the same time from the same script.

**Comment by tianon** [comment]
I'd also add that I'd be happy to do the changes I've proposed here myself if you'd rather I carry this through. :smile:

**Comment by StefanScherer** [comment]
armhf is fine for me, and I'll change the directory tree to "builder/deb/ARCH". 

While rebasing I have to add the armhf/debian:jessie image as well while PR #19076 isn't merged to make the build work. I also add the `bundle .detect-daemon-osarch` in the `build_deb` script as some envs are needed there to select the arch.

File: Dockerfile.armhf [hunk 3]
```diff
@@ -171,9 +171,9 @@ RUN ln -sv $PWD/contrib/completion/bash/docker /etc/bash_completion.d/docker
 # Get useful and necessary Hub images so we can "docker load" locally instead of pulling
 COPY contrib/download-frozen-image-v2.sh /go/src/github.com/docker/docker/contrib/
 RUN ./contrib/download-frozen-image-v2.sh /docker-frozen-images \
-	hypriot/armhf-busybox:latest@sha256:b0fc94dac9793ce3c35607b15012b4c7deca300963a7cc38ab440189ec81e2e7 \
-	hypriot/armhf-hello-world:latest@sha256:b618ec0cc3acf683e8d77ad6c5ec81546cddde2036eda9a78f628effdeca74cd \
-	hypriot/armhf-unshare:latest@sha256:8fede091760d2fb8b2d14cedffdd681c4575b02b1abeeb18dd79b754c62327db
+	armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4 \
```
Comment (StefanScherer) [inline Dockerfile.armhf]: We took @tianon's new official armhf images from PR #19076 as it is easier to introduce the "debian:jessie" for ARM to fix the build.

File: hack/make/.detect-daemon-osarch [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (StefanScherer) [inline hack/make/.detect-daemon-osarch]: This is also from @tianon's PR #19076 as we need these variables for `build-deb` and `build-rpm` scripts to select the correct osarch. We will rebase after 19076 get merged ;-)

**Comment by StefanScherer** [comment]
Rebase done. It depends on PR #19217 to fix the problem building with the `seccomp` build tag. It also embraces the PR #19076 :heart:

**Comment by StefanScherer** [comment]
Oops, all failing builds have `java.lang.NullPointerException: no workspace from node hudson.plugins.ec2.EC2OndemandSlave`. Any chance to restart the builds?

**Comment by StefanScherer** [comment]
Thanks for restarting. The arm build will fail as cherry-pick PR #19217 is still needed for seccomp.

**Comment by justincormack** [comment]
#19217 is now merged, but there are some other errors.

**Comment by StefanScherer** [comment]
Yes, the arm build shows the result after the merge of #19217. I suggest to handle the remaining integration-cli errors in a different PR as this one is for the deb build.
We'll send another PR for that. ;-)

**Comment by jfrazelle** [comment]
i think this just needs a rebase :)

File: contrib/builder/deb/arm/debian-jessie/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FROM armhf/debian:jessie
```
Comment (tianon) [inline contrib/builder/deb/arm/debian-jessie/Dockerfile]: This file still needs to swap from `contrib/builder/deb/arm/debian-jessie/Dockerfile` to `contrib/builder/deb/armhf/debian-jessie/Dockerfile`

File: contrib/reprepro/suites.sh [hunk 6]
```diff
@@ -5,7 +5,7 @@ cd "$(dirname "$BASH_SOURCE")/../.."
 
 targets_from() {
        git fetch -q https://github.com/docker/docker.git "$1"
-       git ls-tree -r --name-only "$(git rev-parse FETCH_HEAD)" contrib/builder/deb | grep '/Dockerfile$' | sed -r 's!^contrib/builder/deb/|-debootstrap|/Dockerfile$!!g'
+       git ls-tree -r --name-only "$(git rev-parse FETCH_HEAD)" contrib/builder/deb/${DOCKER_ENGINE_GOARCH:="amd64"} | grep '/Dockerfile$' | sed -r 's!^contrib/builder/deb/|-debootstrap|/Dockerfile$!!g'
```
Comment (tianon) [inline contrib/reprepro/suites.sh]: Shouldn't this actually use `*` (shell glob) instead of the arch of the engine we're currently running against to make sure we have the _full_ list of possible suites?

Comment (tianon) [inline contrib/reprepro/suites.sh]: I think we probably need to also adapt the `sed` to remove the `arch` component from the path, right?

File: hack/make/.build-rpm/docker-engine.spec [hunk 7]
```diff
@@ -22,7 +22,7 @@ Packager: Docker <support@docker.com>
 %endif
 
 # required packages for build
-# most are already in the container (see contrib/builder/rpm/generate.sh)
+# most are already in the container (see contrib/builder/rpm/(amd64|arm)/generate.sh)
```
Comment (tianon) [inline hack/make/.build-rpm/docker-engine.spec]: Maybe just `ARCH` here so we don't have to update this comment each time we add another?

File: hack/make/build-deb [hunk 8]
```diff
@@ -38,7 +39,7 @@ set -e
 	# TODO decide if it's worth getting go-md2man in _each_ builder environment to avoid this
 
 	# TODO add a configurable knob for _which_ debs to build so we don't have to modify the file or build all of them every time we need to test
-	for dir in contrib/builder/deb/*/; do
+	for dir in contrib/builder/deb/${DOCKER_ENGINE_GOARCH:="amd64"}/*/; do
```
Comment (tianon) [inline hack/make/build-deb]: `:=` modifies the variable -- this should probably just be `:-` instead (or we should just update `.detect-daemon-osarch` to include `: ${DOCKER_ENGINE_GOARCH:=amd64}` at the end to ensure it's always set)

File: contrib/builder/deb/arm/debian-jessie/Dockerfile (hunk 5 above)
Comment (StefanScherer) [inline contrib/builder/deb/arm/debian-jessie/Dockerfile]: @tianon This folder deb/arm/ is named by the values from Go / DOCKER_ENGINE_GOARCH. And there it is called "arm" for 32bit and "aarch64" for 64bit (AFAIK).
Introducing "armhf" here we would need another mapping / variable for that.

Comment (tianon) [inline contrib/builder/deb/arm/debian-jessie/Dockerfile]: We probably should move the `if` chain from the `Makefile` out so we have a
more standard "prefix" or "arch designation" (or both, really) and then it
could be used in more places and when we finally hear back from ARM about
what they'd officially like us to call each of these we can change it in
one main place.

Comment (tianon) [inline contrib/builder/deb/arm/debian-jessie/Dockerfile]: More specifically, I mean move it to the detect-os-arch script as a
"case" instead.

File: contrib/reprepro/suites.sh (hunk 6 above)
Comment (StefanScherer) [inline contrib/reprepro/suites.sh]: The `sed` worked correctly building this PR. Could it be that it worked because we are in a branch and FETCH_HEAD commit id still has the old structure?

The build-deb.sh still only generates the amd64 Dockerfiles. The arm/debian-jessie Dockerfile is built manually at the moment. When we introduce more than one ARM OS platform we will update the build-deb.sh in another PR. That was our plan to get at least the new folder structure in it. Rebasing moving folders is no fun. ;-)

File: hack/make/build-deb (hunk 8 above)
Comment (StefanScherer) [inline hack/make/build-deb]: Now as the `.detect-daemon-osarch` is available and the `DOCKER_ENGINE_GOARCH` is always set we could remove the default values here. Sounds good.

File: hack/make/.build-rpm/docker-engine.spec (hunk 7 above)
Comment (StefanScherer) [inline hack/make/.build-rpm/docker-engine.spec]: OK.

File: contrib/reprepro/suites.sh (hunk 6 above)
Comment (tianon) [inline contrib/reprepro/suites.sh]: Yeah, both the branches this is fetching from still have the old structure,
so this sed will definitely need to be a little bit more complicated.

**Comment by StefanScherer** [comment]
Another rebase as of #19377 - let's see how this works together for multi arch.

**Comment by thaJeztah** [comment]
@StefanScherer :cry: needs a rebase again already

**Comment by StefanScherer** [comment]
@thaJeztah my last comment was only a notice that I have to do that. ;-)  Now it's rebased.

**Comment by thaJeztah** [comment]
experimental timed out; restarting

**Comment by StefanScherer** [comment]
arm is :green_heart:

**Comment by Govinda-Fichtner** [comment]
Finally.
Am 22.01.2016 08:25 schrieb "Stefan Scherer" <notifications@github.com>:

**Comment by thaJeztah** [comment]
ping @tianon @jfrazelle it's greeeeeeeen!!

**Comment by jfrazelle** [comment]
ok im going to run some test builds then will merge :)

File: contrib/builder/deb/arm/debian-jessie/Dockerfile (hunk 5 above)
Comment (jfrazelle) [inline contrib/builder/deb/arm/debian-jessie/Dockerfile]: also this should be taken into account :)

**Comment by thaJeztah** [comment]
oh, bummer; needs a rebase again :cry:

**Comment by thaJeztah** [comment]
Got one failure on gccgo, not sure it's related;

```
15:20:38 
15:20:38 ----------------------------------------------------------------------
15:20:38 FAIL: docker_cli_network_unix_test.go:995: TestDockerNetworkHostModeUngracefulDaemonRestart.pN59_github_com_docker_docker_integration_cli.DockerNetworkSuite
15:20:38 
15:20:38 [d1326000] waiting for daemon to start
15:20:38 [d1326000] daemon started
15:20:38 [d1326000] exiting daemon
15:20:38 [d1326000] waiting for daemon to start
15:20:38 [d1326000] daemon started
15:20:38 docker_cli_network_unix_test.go:1016:
15:20:38     c.Assert(err, checker.IsNil)
15:20:38 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc2081439a0)} ("exit status 1")
15:20:38 
15:20:38 [d1326000] exiting daemon
15:20:40 
15:20:40 ----------------------------------------------------------------------
```

**Comment by jfrazelle** [comment]
I keep getting stuck here on every single arm server ive tried :(, is this a hub issue? 
```
Step 34 : RUN ./contrib/download-frozen-image-v2.sh /docker-frozen-images 	armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb 	armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4 	armhf/debian:jessie@sha256:4a2187483f04a84f9830910fe3581d69b3c985cc045d9f01d8e2f3795b28107b 	armhf/hello-world:latest@sha256:161dcecea0225975b2ad5f768058212c1e0d39e8211098666ffa1ac74cfb7791
 ---> Running in ffee715d1d6d
error: /v2/armhf/buildpack-deps/manifests/sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb returned something unexpected:
  <html><body><h1>408 Request Time-out</h1>
Your browser didn't send a complete request in time.
</body></html>
```

**Comment by StefanScherer** [comment]
I got this about two weeks ago. I'll restart a regression test with master branch with a clean Jenkins slave and see how it works in our CI. Our last jenkins builds used cached image layers.

**Comment by StefanScherer** [comment]
@thaJeztah Isn't TestDockerNetworkHostModeUngracefulDaemonRestart flaky? https://github.com/docker/docker/issues/19368

**Comment by thaJeztah** [comment]
@StefanScherer oh, you're right! I was in a hurry, so didn't check. Let me restart CI

**Comment by StefanScherer** [comment]
Hm, `./contrib/download-frozen-image-v2.sh` works fine here, took less than two minutes:

```
18:19:19 Removing intermediate container 459fc2166764
18:19:19 Step 34 : RUN ./contrib/download-frozen-image-v2.sh /docker-frozen-images 	armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb 	armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4 	armhf/debian:jessie@sha256:4a2187483f04a84f9830910fe3581d69b3c985cc045d9f01d8e2f3795b28107b 	armhf/hello-world:latest@sha256:161dcecea0225975b2ad5f768058212c1e0d39e8211098666ffa1ac74cfb7791
18:19:19  ---> Running in b23c5513f6da
18:19:22 Downloading 'armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb' (5 layers)...
18:19:54 
...
18:20:55 Download of images into '/docker-frozen-images' complete.
18:20:55 Use something like the following to load the result into a Docker daemon:
18:20:55   tar -cC '/docker-frozen-images' . | docker load
18:21:02  ---> c65aefc047cc
18:21:03 Removing intermediate container b23c5513f6da
18:21:03 Step 35 : RUN set -x 	&& export GOPATH="$(mktemp -d)" 	&& git clone --depth 1 -b v1.0.4 
```

**Comment by thaJeztah** [comment]
And.. hitting https://github.com/docker/docker/issues/19368 again :disappointed:

**Comment by jfrazelle** [comment]
@StefanScherer I can't get it to work, is there a proxy involved?

**Comment by StefanScherer** [comment]
@jfrazelle No proxy involved. I'm setting up a new Scaleway machine and run a complete fresh build. Let's see what happens there.

**Comment by StefanScherer** [comment]
rebased once again ;-)

**Comment by StefanScherer** [comment]
Running the command locally on my Mac works, running it directly on a Scaleway server also works fine.

Testing it with this little Dockerfile.test on a Scaleway machine also works:

```
FROM armhf/ubuntu:trusty
RUN apt-get update && apt-get install -y curl jq
COPY . /code
WORKDIR /code
RUN ./contrib/download-frozen-image-v2.sh ./tmp/ armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb     armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4    armhf/debian:jessie@sha256:4a2187483f04a84f9830910fe3581d69b3c985cc045d9f01d8e2f3795b28107b     armhf/hello-world:latest@sha256:161dcecea0225975b2ad5f768058212c1e0d39e8211098666ffa1ac74cfb7791
CMD [ "bash" ]
```

```bash
docker build -t test -f Dockerfile.test .
```

Only thing is that `go` isn't installed, but the ìf`command in line 60 can be ignored for this test.
And the fresh Scaleway Jenkins slave also has built the image and is running the integration-cli tests at the moment.

**Comment by icecrime** [comment]
@jfrazelle Can you please check again?

**Comment by StefanScherer** [comment]
Setup of a new Scaleway instance is done in three steps:

```
scw create --name=scw-07 Docker_1_9_0
# wait a minute
scw start scw-07
# wait a minute
ssh root@$(scw inspect server:scw-07 | jq -r .[].public_ip.address) 'bash -s' < install-jenkins-swarm-slave.sh
```

installs openjdk-7-jre, make, fuse, the jenkins swarm client and an init script.

**Comment by StefanScherer** [comment]
This time gccgo failed with `TestSearchCmdOptions`.

**Comment by jfrazelle** [comment]
I can consistently not hit hub, ping @justincormack who has seen the same
thing @avsm too

**Comment by StefanScherer** [comment]
Ping @moul any ideas what might happen from a scaleway machine connecting Docker Hub?

**Comment by avsm** [comment]
I'm also seeing really long timeouts when trying to connect to the Hub from a scaleway.  There seems to be some sort of proxy involved...

**Comment by jfrazelle** [comment]
ping index.docker.io just hangs for me

**Comment by Govinda-Fichtner** [comment]
@jfrazelle for me, too:
```
$ ping index.docker.io
PING us-east-1-elbio-rm5bon1qaeo4-623296237.us-east-1.elb.amazonaws.com (54.164.250.255): 56 data bytes
Request timeout for icmp_seq 0
Request timeout for icmp_seq 1
Request timeout for icmp_seq 2
Request timeout for icmp_seq 3
Request timeout for icmp_seq 4
Request timeout for icmp_seq 5
Request timeout for icmp_seq 6
Request timeout for icmp_seq 7
```

**Comment by jfrazelle** [comment]
ok so this is my bad, ping is turned off for that domain... so sorry but
the problem still exists that was just a bad example

**Comment by StefanScherer** [comment]
I checked my old jenkins logs and found out that my 408 request timeout problems only occured on one scaleway machine created about two weeks ago. The new machine from yesterday just works out of the box.
So either parts of the datacenter have problems or the machine image had a problems. 
@jfrazelle Does a newly created scaleway machine work for you or is there the same problem?

**Comment by jfrazelle** [comment]
i will try

**Comment by jfrazelle** [comment]
It's still hanging :/ I cant even run the script on the host...

**Comment by djmaze** [comment]
@StefanScherer @jfrazelle This is probably related to the (high) MTU Scaleway sets on some of their servers. Try disabling path MTU discovery or lower the MTU on the eth device. See this thread for more information: https://github.com/scaleway-community/scaleway-docker/issues/30#issuecomment-141981774

**Comment by aimxhaisse** [comment]
Hello from Scaleway,

As highlighted by @djmaze, the connectivity issues you sometimes encounter on Scaleway nodes is related to the MTU being at 9K. Most of our infrastructure is at 1.5k but we have a few spots left at 9K (we are progressively deprecating 9K in favor of 1.5K but this is a move we can't do everywhere without impacting other nodes, so it takes time).

The issue is that the PMTU discovery does not properly work because of a hop on our side, we have the following workaround (which we are going to temporarily integrate in our Docker image):

    echo 1 > /proc/sys/net/ipv4/ip_no_pmtu_disc

Basically, it disables PMTU discovery when receiving an ICMP error on a too large packet, and fallbacks at 552 when a 9K MTU is not possible.

**Comment by StefanScherer** [comment]
@aimxhaisse @djmaze Thank you!

**Comment by moul** [comment]
We just released a [new version](http://devhub.scaleway.com/#/images/docker:latest) of the Docker image on Scaleway containing the last fix:

```console
$ scw run docker
[...]
root@agitated-hamilton:~# sysctl net.ipv4.ip_no_pmtu_disc
net.ipv4.ip_no_pmtu_disc = 1
```

**Comment by thaJeztah** [comment]
Thanks @moul! @jfrazelle ^^

**Comment by jfrazelle** [comment]
Cool I will try it out

File: hack/make/release-deb [hunk 9]
```diff
@@ -122,7 +122,7 @@ apt-ftparchive clean "$APTDIR/conf/apt-ftparchive.conf"
 # run the apt-ftparchive commands so we can have pinning
 apt-ftparchive generate "$APTDIR/conf/apt-ftparchive.conf"
 
-for dir in contrib/builder/deb/*/; do
+for dir in contrib/builder/deb/${DOCKER_ENGINE_GOARCH}/*/; do
```
Comment (jfrazelle) [inline hack/make/release-deb]: ok so considering the debs will all be released from a amd64 machine we need a better way to do this. I am happy to make the patch in a different PR because I know what I have in mind

Comment (jfrazelle) [inline hack/make/release-deb]: I don't think it should be a blocker im more putting this here so you are aware

**Comment by jfrazelle** [comment]
can we change line 56 of build-deb so the destination is something like `"$DEST/$version/$arch"` it will make releasing easier

**Comment by jfrazelle** [comment]
also arm needs to be added to https://github.com/docker/docker/pull/18176/files#diff-c6c6bb0d2e164817a23d98d07109b25aL25 (it wont let me comment on the line) 
then in release-deb we can iterate through the arches and check if there are packages in `bundles/$VERSION/build-deb/$arch/$suite` if that makes sense, its better since we will release from amd64 probably 100% of the time

**Comment by jfrazelle** [comment]
also I spun up a new instance and its still timing out on that one line, I would be happy to give you ssh access want to email me at jess@docker.com

**Comment by tianon** [comment]
Don't the output filenames already have arch in them?

**Comment by jfrazelle** [comment]
oh true but I was thinking if we wanted to release amd64 and arm at the same time

**Comment by jfrazelle** [comment]
like combine the bundles, but maybe that is counter intuitive

**Comment by tianon** [comment]
Right, wouldn't we just put all the files in the same place and point the
release script at the directory?

**Comment by jfrazelle** [comment]
yeah that!

**Comment by jfrazelle** [comment]
ok so i guess its fine now and we can fix `release-deb to iterate through dirs in bundles/version/build-deb?

**Comment by tianon** [comment]
Just built again to verify filenames, and here's what we've got right now:
```console
$ ls bundles/latest/build-deb/debian-jessie/
docker-engine_1.10.0~dev~git20160127.162101.0.ae14e6c-0~jessie_amd64.changes
docker-engine_1.10.0~dev~git20160127.162101.0.ae14e6c-0~jessie_amd64.deb
docker-engine_1.10.0~dev~git20160127.162101.0.ae14e6c-0~jessie.dsc
docker-engine_1.10.0~dev~git20160127.162101.0.ae14e6c-0~jessie.tar.gz
Dockerfile.build
```

Those files that aren't tagged with the arch already should be basically the same on each arch (sans mtime metadata, likely); in some future universe we might consider swapping all of this to generate one deterministic `.dsc` file and use that to build all the arches (like the real archive buildds do), but I think this is perfect for now (since the tooling for doing that sort of thing is anything but friendly).

Also, IMO we should hold off on actually building and releasing these artifacts until after we're ready to commit fully to multiarch support (ie, Hub, Registry, Engine, etc all handle multiarch in a reasonable and meaningful way).

**Comment by jfrazelle** [comment]
ok sweet

**Comment by stefanschererseal** [comment]
The files from an arm build:
```
# ls bundles/1.10.0-dev/build-deb/debian-jessie/
docker-engine_1.10.0~dev~git20151123.190043.0.02a6ecb-0~jessie_armhf.changes
docker-engine_1.10.0~dev~git20151123.190043.0.02a6ecb-0~jessie_armhf.deb
docker-engine_1.10.0~dev~git20151123.190043.0.02a6ecb-0~jessie.dsc
docker-engine_1.10.0~dev~git20151123.190043.0.02a6ecb-0~jessie.tar.gz
Dockerfile.build
```

So changes for `"$DEST/$arch/$version"` and `bundles/$VERSION/build-deb/$arch/$suite` should be done in this PR, right?

**Comment by tianon** [comment]
> So changes for `"$DEST/$arch/$version"` and `bundles/$VERSION/build-deb/$arch/$suite` should be done in this PR, right?

Naw, since the `.deb` files themselves include the arch, I don't think we need to differentiate these files further at all (that's what I was discussing with @jfrazelle).

**Comment by jfrazelle** [comment]
no sorry its perfect the way it is, i just made this super confusing for no
reason ;)

**Comment by moul** [comment]
Here is a fix for the MTU issue when building Docker within Docker (dind): https://github.com/scaleway-community/scaleway-docker/pull/51

We will release a new version of the image tomorrow with the fix built-in

**Comment by jfrazelle** [comment]
You are the best!!!!

**Comment by jfrazelle** [comment]
so we are only building arm on debian jessie then?

**Comment by StefanScherer** [comment]
Yes, just one DEB package for ARM in this PR. This PR introduces the changes for the new directory structure.

**Comment by jfrazelle** [comment]
so I think we are all good once this is addressed :) https://github.com/docker/docker/pull/18176#discussion_r50197993

**Comment by StefanScherer** [comment]
Ok, will do. 
Should I also rename the Intel version `Dockerfile` to `Dockerfile.amd64` to be consequent? 😄

File: hack/make/.detect-daemon-osarch [hunk 10]
```diff
@@ -16,3 +17,18 @@ export DOCKER_CLIENT_OSARCH="$(docker version | awk '
 	$1 == "Server:" { client = 0; next }
 	client && $1 == "OS/Arch:" { print $2 }
 ')"
+
+# Retrieve the architecture used in contrib/builder/(deb|rpm)/$PACKAGE_ARCH/
+PACKAGE_ARCH="amd64"
+case "$DOCKER_ENGINE_OSARCH" in
+	linux/arm)
+		PACKAGE_ARCH='armhf'
```
Comment (StefanScherer) [inline hack/make/.detect-daemon-osarch]: Added a variable `$PACKAGE_ARCH` for the `contrib/builder/(deb|rpm)/$PACKAGE_ARCH/` sub directories as we want eg. `armhf` instead of `$DOCKER_ENGINE_GOARCH` which is `arm`.

**Comment by jfrazelle** [comment]
LGTM, ping @tianon

**Comment by StefanScherer** [comment]
The ARM DEB package works fine on a Scaleway machine.

Upgrade path for the brave admins:

```
systemctl stop docker
apt-get remove -y docker-hypriot
rm /etc/systemd/system/docker.service
wget http-url-to-docker-engine_1.10.0%7Edev%7Egit20151123.190043.0.7b31ed4-0%7Ejessie_armhf.deb
dpkg -i docker-engine_1.10.0%7Edev%7Egit20151123.190043.0.7b31ed4-0%7Ejessie_armhf.deb
```

Overwrite the `/etc/default/docker` file, but it is no longer needed.
Removing the `/etc/systemd/system/docker.service` is required as it contains the deprecated `-d` option.

```
root@scw-04:~# docker version
Client:
 Version:      1.10.0-dev
 API version:  1.23
 Go version:   go1.4.3
 Git commit:   7b31ed4
 Built:        Fri Jan 29 00:11:09 2016
 OS/Arch:      linux/arm

Server:
 Version:      1.10.0-dev
 API version:  1.23
 Go version:   go1.4.3
 Git commit:   7b31ed4
 Built:        Fri Jan 29 00:11:09 2016
 OS/Arch:      linux/arm
```

**Comment by moul** [comment]
`Go version: go1.4.3`, is this something important to update soon ?

**Comment by StefanScherer** [comment]
#18125 seems to be the next big challenge. Will try go 1.6rc1 soon.

**Comment by thaJeztah** [comment]
gccgi: `Build timed out (after 180 minutes)` :cry:

**Comment by thaJeztah** [comment]
\o/

**Comment by StefanScherer** [comment]
:rocket:

**Comment by DieterReuter** [comment]
:tophat:

**Comment by Govinda-Fichtner** [comment]
Finally :racehorse: over the finishing line.
